core/domain/question_services.py

```diff
@@ -209,21 +209,32 @@ def get_questions_and_skill_descriptions_by_skill_ids(
             questions are to be returned. This value should be urlsafe.
 
     Returns:
-        list(Question), list(str), str. The list of questions and the
+        list(Question), list(list(str)), str. The list of questions and the
             corresponding linked skill descriptions which are linked to the
             given skill ids and the next cursor value to be used for the next
             batch of questions (or None if no more pages are left). The returned
             next cursor value is urlsafe.
     """
-    question_skill_link_models, skill_descriptions, next_cursor = (
-        question_models.QuestionSkillLinkModel.get_question_skill_links_and_skill_descriptions( #pylint: disable=line-too-long
+    question_skill_link_models, next_cursor = (
+        question_models.QuestionSkillLinkModel.get_question_skill_links_by_skill_ids( #pylint: disable=line-too-long
             question_count, skill_ids, start_cursor))
-    question_skill_links = [
-        get_question_skill_link_from_model(model, skill_descriptions[index])
-        for index, model in enumerate(question_skill_link_models)]
-    question_ids = [obj.question_id for obj in question_skill_links]
+    question_ids = []
+    grouped_skill_ids = []
+    grouped_skill_descriptions = []
+    for question_skill_link in question_skill_link_models:
+        if question_skill_link.question_id not in question_ids:
+            question_ids.append(question_skill_link.question_id)
+            grouped_skill_ids.append([question_skill_link.skill_id])
+        else:
+            grouped_skill_ids[-1].append(question_skill_link.skill_id)
```

Comment:
This assumes that  all question skill links with a question id are grouped together, right? Is that the case?

Reply:
Let's say question 1 has two linked skills, and question 2 has one linked skill, then question_ids will be ['question_1', 'question_2'], and grouped_skill_ids will be [['skill_1', 'skill_2'], ['skill_3']]